runtime/util/thrinfo.c

```diff
@@ -420,8 +420,9 @@ getVMThreadStateHelper(J9VMThread *targetThread,
 								lockOwnerObject = J9OBJECT_FROM_FJ9OBJECT(targetThread->javaVM, fobj);
 							}
 #else
+							/* Simplify the macro usage by using the _VM version so we do not need to pass in the current thread */
 							lockOwnerObject =
-								J9VMJAVAUTILCONCURRENTLOCKSABSTRACTOWNABLESYNCHRONIZER_EXCLUSIVEOWNERTHREAD(targetThread, lockObject);
+								J9VMJAVAUTILCONCURRENTLOCKSABSTRACTOWNABLESYNCHRONIZER_EXCLUSIVEOWNERTHREAD_VM(targetThread->javaVM, lockObject);
```

Comment:
This change does not make sense to me... The modified code is inside `if(targetThread) {` so I do not see how this change is required. Can you provide more details please?

Reply:
First argument should be the calling thread. It has no association with lockObject. First argument is there just to be able to fetch native resources/info if needed - it's not used in any semantics with what this this macro does (everything is inferred from the second argument).

One piece of info that first argument is needed for is offset, and it would work with any thread provided, as Dan already observed.
The problematic case is if the info/resource is thread specific - for example, it  could be various thread local buffers, like (in GC world) allocation/copy/Remember Set etc. buffers. 

Specifically, the problem is observed for Read Barrier for Evacuating GC, where the object accessed (a child of lockObject) is copied to targetThread local copy buffer, which is not the thread executing the macro.

By using _VM API, the current thread is extracted by an OS call.

Reply:
@charliegracie While running concurrent scavenge on X86, we had seen this GC assert: ``** ASSERTION FAILED ** at StandardAccessBarrier.cpp:675: ((false && (_extensions->scavenger->isMutatorThreadInSyncWithCycle(env))))``. 
Aleks looked into this assert and believe the problem was due to the macro `J9VMJAVAUTILCONCURRENTLOCKSABSTRACTOWNABLESYNCHRONIZER_EXCLUSIVEOWNERTHREAD` where the first argument `targetThread` passed in is not the current thread, where the original macro (or the rdbar code) expects it to be. So the proposed fix will pass in the current thread to the rdbar.

Reply:
ok after inspecting the code the J9VMThread passed to this function is almost never the current thread.  That means that we need to figure out the current J9VMThread for the barrier call.. The code change achieves this by pass the JavaVM as the first parameter and then getting the currentVMThread() from the JavaVM.

Reply:
I'm concerned that this is going to be a game of whack-a-mole.  Shouldn't line 431, which uses the `J9VMJAVALANGTHREAD_THREADREF(targetThread, lockOwnerObject);` macro get the same treatment?

Reply:
I raised that concern already (https://github.com/eclipse/openj9/pull/3256#discussion_r224896433). If you want we can fix it. Alternatively, we can wait till we do more comprehensive testing and potentially more spots that we know for sure are broken (for some reason, this specific spot is does not appear to be problematic, could be pure luck).

Reply:
Yes, let's fix it now!  I would like all instances of the problem in this function to be addressed at the same time.

Reply:
@DanHeidinga Fixed both instances.

Reply:
I just checked if  J9VMJAVALANGTHREAD_THREADREF and there is a subtle detail on what kind of object access API they rely (J9OBJECT_ADDRESS_LOAD vs J9OBJECT_OBJECT_LOAD). The former for Concurrent Scavenger (and true for any collector, so far) does not have any barrier (it just reads the value). While this explains if why CS did not exhibit a problem, it would be still cleaner to fix this spot, too.